frappe/api.py

```diff
@@ -143,48 +152,82 @@ def handle():
-		from frappe.integrations.oauth2 import get_oauth_server
-		token = authorization_header[1]
-		req = frappe.request
-		parsed_url = urlparse(req.url)
-		access_token = {"access_token": token}
-		uri = parsed_url.scheme + "://" + parsed_url.netloc + parsed_url.path + "?" + urlencode(access_token)
-		http_method = req.method
-		body = req.get_data()
-		headers = req.headers
+	token = authorization_header[1]
+	req = frappe.request
+	parsed_url = urlparse(req.url)
+	access_token = {"access_token": token}
+	uri = parsed_url.scheme + "://" + parsed_url.netloc + parsed_url.path + "?" + urlencode(access_token)
+	http_method = req.method
+	body = req.get_data()
+	headers = req.headers
 
+	try:
 		required_scopes = frappe.db.get_value("OAuth Bearer Token", token, "scopes").split(get_url_delimiter())
+	except AttributeError:
+		frappe.throw(_("Invalid Bearer token, please provide a valid access token with prefix 'Bearer'."), InvalidAuthorizationToken)
 
-		valid, _oauthlib_request = get_oauth_server().verify_request(uri, http_method, body, headers, required_scopes)
+	valid, oauthlib_request = get_oauth_server().verify_request(uri, http_method, body, headers, required_scopes)
 
-		if valid:
-			frappe.set_user(frappe.db.get_value("OAuth Bearer Token", token, "user"))
-			frappe.local.form_dict = form_dict
+	if valid:
+		frappe.set_user(frappe.db.get_value("OAuth Bearer Token", token, "user"))
+		frappe.local.form_dict = form_dict
 
 
-def validate_auth_via_api_keys():
+def validate_auth_via_api_keys(authorization_header):
 	"""
-	authentication using api key and api secret
+	Authenticate request using API keys and set session user
 
-	set user
+	Args:
+		authorization_header (list of str): The 'Authorization' header containing the prefix and token
 	"""
+
 	try:
-		authorization_header = frappe.get_request_header("Authorization", "").split(" ")
+		auth_type, auth_token = authorization_header
 		authorization_source = frappe.get_request_header("Frappe-Authorization-Source")
-		if authorization_header and authorization_header[0] == 'Basic':
-			token = frappe.safe_decode(base64.b64decode(authorization_header[1])).split(":")
-			validate_api_key_secret(token[0], token[1], authorization_source)
-		elif authorization_header and authorization_header[0] == 'token':
-			token = authorization_header[1].split(":")
-			validate_api_key_secret(token[0], token[1], authorization_source)
-	except Exception as e:
-		raise e
+		if auth_type.lower() == 'basic':
+			api_key, api_secret = frappe.safe_decode(base64.b64decode(auth_token)).split(":")
+			validate_api_key_secret(api_key, api_secret, authorization_source)
+		elif auth_type.lower() == 'token':
+			api_key, api_secret = auth_token.split(":")
+			validate_api_key_secret(api_key, api_secret, authorization_source)
+	except (AttributeError, TypeError, ValueError):
+		frappe.throw(_("Invalid token, please provide a valid token with prefix 'Basic' or 'Token'."), InvalidAuthorizationToken)
+	except binascii.Error:
```

Comment:
shouldn't this come above the other exceptions?

Reply:
hm, deepsource says so too: https://deepsource.io/gh/frappe/frappe/run/c434b531-d0d6-43fc-b0fd-a80a27b0f3d3/python/PYL-E0701